### ISSUE TITLE
Compilation problem when using -dist / -cache option with Boost

### ISSUE THREAD

**Issue by esavard** [fastbuild/fastbuild#181]
Hi,

I have a strange problem. My environment is VS2013 on Windows 10 using FBuild 0.92. When I use the command `fbuild -dist -cache -clean x64-debug`, the compilation fail with this error:

```
PROBLEM: C:\Users\esavard\Work\builds\x64-Debug\Obj\TireInspection\src\SidewallMap.obj
PROBLEM: C:\Users\esavard\AppData\Local\Temp\.fbuild.tmp\0x00000000\core_1004\SidewallMap.obj
SidewallMap.cpp
C:\Users\esavard\Work\lib\x64\include\boost/icl/detail/element_iterator.hpp(116) : error C2065: 'Intervalconst' : undeclared identifier
C:\Users\esavard\Work\lib\x64\include\boost/icl/detail/element_iterator.hpp(116) : error C2923: 'std::pair' : 'Intervalconst' is not a valid template type argument for parameter '_Ty1'
C:\Users\esavard\Work\lib\x64\include\boost/icl/detail/element_iterator.hpp(117) : error C3203: 'pair' : unspecialized class template can't be used as a template argument for template parameter 'Interval', expected a real type
C:\Users\esavard\Work\lib\x64\include\boost/icl/detail/element_iterator.hpp(115) : error C3211: 'boost::icl::elemental<int>' : explicit specialization is using partial specialization syntax, use template <> instead
        C:\Users\esavard\Work\lib\x64\include\boost/icl/detail/element_iterator.hpp(117) : see declaration of 'boost::icl::elemental<int>'
C:\Users\esavard\Work\lib\x64\include\boost/icl/detail/element_iterator.hpp(217) : error C2065: 'Intervalconst' : undeclared identifier
C:\Users\esavard\Work\lib\x64\include\boost/icl/detail/element_iterator.hpp(217) : error C2923: 'std::pair' : 'Intervalconst' is not a valid template type argument for parameter '_Ty1'
C:\Users\esavard\Work\lib\x64\include\boost/icl/detail/element_iterator.hpp(218) : error C3203: 'pair' : unspecialized class template can't be used as a template argument for template parameter 'Interval', expected a real type
C:\Users\esavard\Work\lib\x64\include\boost/icl/detail/element_iterator.hpp(239) : error C2764: 'CodomainT' : template parameter not used or deducible in partial specialization 'boost::icl::segment_adapter<SegmentIteratorT,int>'
C:\Users\esavard\Work\lib\x64\include\boost/icl/detail/element_iterator.hpp(239) : error C2764: 'Interval' : template parameter not used or deducible in partial specialization 'boost::icl::segment_adapter<SegmentIteratorT,int>'
Failed to build Object (error 0x2) 'C:\Users\esavard\AppData\Local\Temp\.fbuild.tmp\0x00000000\core_1004\SidewallMap.obj'
6> Obj: C:\Users\esavard\Work\builds\x64-Debug\Obj\3d\CalibrateCamera2d.obj
-> Obj: C:\Users\esavard\Work\builds\x64-Debug\Obj\ImgProc\FindLocalMinMaxSubPixAdaptative.obj <REMOTE: NUMSRVCI01>
FBuild: Error: BUILD FAILED: x64
Time: 9.460s
```

I've tried to erase the AppData\Local\Temp\.fbuild.tmp folder and restart the compilation using the same options but I got the same error. Compiling using only either `-dist -clean` or `-cache -clean` options yield the same results.

If I compile locally (ie: using `fbuild -clean x64-debug`), the compilation complete without any errors...

Any idea what could be the cause of this problem and how to work around it?

Thanks.

**Comment by haferburg** [comment]
`Intervalconst` might actually be `Interval const`. See the remarks on https://msdn.microsoft.com/en-us/library/3xkfswhy.aspx .

**Comment by ffulin** [comment]
I believe haferburg is correct. The MSVC pre-processor has inconsistent treatment of macro expansion when using the preprocessor (which FASTBuild uses for caching and distribution).

I took a quick look at the boost header in question, and I found this pattern on a few lines:
```C++
ICL_INTERVAL_TYPE(Interval,DomainT,Compare)const
```
The fix in this case is to insert a space between the closing ) of the macro and the const.  This is a reasonable fix that could be pulled into boost to avoid future maintenance hassles.

**Comment by esavard** [comment]
Thanks, it worked.

**Comment by ffulin** [comment]
Excellent.  Do you think you will try to push the fix upstream to boost?

**Comment by esavard** [comment]
Yes, I will try to get my fix included in boost.  I will post back here when it's done.

**Comment by esavard** [comment]
Seems that there is already a patch that was submitted to boost::icl months ago...

https://svn.boost.org/trac/boost/ticket/12375

**Comment by ffulin** [comment]
Thanks for the follow up. I guess we can leave this ticket open while we wait for that to be integrated into Boost.

**Comment by esavard** [comment]
We can mark this issue as "fixed", boost issue [boostorg/icl#7](https://github.com/boostorg/icl/pull/7) was fixed.

**Comment by ffulin** [comment]
@esavard Great stuff. Thanks for letting me know :)
